plugins/Model_Original.py

```diff
@@ -1,68 +1,73 @@
-from keras.optimizers import Adam
-
-from lib.ModelAE import ModelAE, TrainerAE
-from lib.PixelShuffler import PixelShuffler
-
-IMAGE_SHAPE = (64, 64, 3)
-ENCODER_DIM = 1024
-
-class Model(ModelAE):
-    def initModel(self):
-        optimizer = Adam(lr=5e-5, beta_1=0.5, beta_2=0.999)
-        x = Input(shape=IMAGE_SHAPE)
-
-        self.autoencoder_A = KerasModel(x, self.decoder_A(self.encoder(x)))
-        self.autoencoder_B = KerasModel(x, self.decoder_B(self.encoder(x)))
-
-        self.autoencoder_A.compile(optimizer=optimizer, loss='mean_absolute_error')
-        self.autoencoder_B.compile(optimizer=optimizer, loss='mean_absolute_error')
-
-    def converter(self, swap):
-        autoencoder = self.autoencoder_B if not swap else self.autoencoder_A 
-        return lambda img: autoencoder.predict(img)
-
-    def conv(self, filters):
-        def block(x):
-            x = Conv2D(filters, kernel_size=5, strides=2, padding='same')(x)
-            x = LeakyReLU(0.1)(x)
-            return x
-        return block
-
-    def upscale(self, filters):
-        def block(x):
-            x = Conv2D(filters * 4, kernel_size=3, padding='same')(x)
-            x = LeakyReLU(0.1)(x)
-            x = PixelShuffler()(x)
-            return x
-        return block
-
-    def Encoder(self):
-        input_ = Input(shape=IMAGE_SHAPE)
-        x = input_
-        x = self.conv(128)(x)
-        x = self.conv(256)(x)
-        x = self.conv(512)(x)
-        x = self.conv(1024)(x)
-        x = Dense(ENCODER_DIM)(Flatten()(x))
-        x = Dense(4 * 4 * 1024)(x)
-        x = Reshape((4, 4, 1024))(x)
-        x = self.upscale(512)(x)
-        return KerasModel(input_, x)
-
-    def Decoder(self):
-        input_ = Input(shape=(8, 8, 512))
-        x = input_
-        x = self.upscale(256)(x)
-        x = self.upscale(128)(x)
-        x = self.upscale(64)(x)
-        x = Conv2D(3, kernel_size=5, padding='same', activation='sigmoid')(x)
-        return KerasModel(input_, x)
-
-class Trainer(TrainerAE):
+# Based on the original https://www.reddit.com/r/deepfakes/ code sample + contribs
+
+from keras.models import Model as KerasModel
+from keras.layers import Input, Dense, Flatten, Reshape
+from keras.layers.advanced_activations import LeakyReLU
+from keras.layers.convolutional import Conv2D
+from keras.optimizers import Adam
+from keras.utils import multi_gpu_model
+from lib.ModelAE import ModelAE, TrainerAE
+from lib.PixelShuffler import PixelShuffler
+
+IMAGE_SHAPE = (64, 64, 3)
+ENCODER_DIM = 1024
+
+class Model(ModelAE):
+    def initModel(self):
+        optimizer = Adam(lr=5e-5, beta_1=0.5, beta_2=0.999)
+        x = Input(shape=IMAGE_SHAPE)
+
+        self.autoencoder_A = KerasModel(x, self.decoder_A(self.encoder(x)))
+        self.autoencoder_B = KerasModel(x, self.decoder_B(self.encoder(x)))
+        self.autoencoder_A = multi_gpu_model( self.autoencoder_A ,2)
+        self.autoencoder_B = multi_gpu_model( self.autoencoder_B ,2)
```

Comment:
According to the [docs](https://keras.io/utils/#multi_gpu_model) the second argument which is a 2 here indicates the number of GPUs to split work between. This shouldn't be hard-coded. What happens if you run this with only 1 GPU? You should add a `--gpus` option to [cli.py](https://github.com/deepfakes/faceswap/blob/232d9313afabc075e28c1a42047b22e0568200e6/lib/cli.py) and use it here.

Reply:
Have you been able to load/save a multi-gpu model? From the Keras docs:
> On model saving:
>
>To save the multi-gpu model, use .save(fname) or .save_weights(fname) with the template model (the argument you passed to multi_gpu_model), rather than the model returned by multi_gpu_model.

So it seems like we still need to maintain a reference to the serial model in order to save.

Reply:
I agree with you, I don't know what happens in a single gpu environment, though I suppose I could disable one of mine to find out. This does work on my system as it splits the load between the two GPU's and uses both the VRAM from the two cards.

Reply:
I was able to save one without problem.  I ran one with this model and it works perfectly.  Saving is already handled by train.py in the correct way.  The only problem with this will be if we start using checkpoints, which by default are not multi GPU compatible.

The models are also still compatible with single GPU, as I've taken a multi GPU model and run it on the same model in single GPU mode and then back in multi GPU, so you don't have to wipe it and restart if you change the number of GPUS available.  Since the nodes and layers are the same, it loads without problem and runs fine.

We should be able to simply programatically find out how many GPUS are available and by default use all of them.

The following link shows how to find a number of usable CUDA GPUS in tensorflow, I'm sure some Keras equivalent exists.
https://stackoverflow.com/questions/38559755/how-to-get-current-available-gpus-in-tensorflow